prov/shm/src/smr_progress.c

```diff
@@ -569,29 +569,47 @@ static int smr_progress_ipc(struct smr_cmd *cmd, enum fi_hmem_iface iface,
 }
 
 static void smr_do_atomic(void *src, void *dst, void *cmp, enum fi_datatype datatype,
-			  enum fi_op op, size_t cnt, uint16_t flags)
+			  enum fi_op op, size_t cnt, uint16_t flags, enum fi_hmem_iface iface, uint64_t device)
 {
 	char tmp_result[SMR_INJECT_SIZE];
 
 	if (ofi_atomic_isswap_op(op)) {
-		ofi_atomic_swap_handler(op, datatype, dst, src, cmp,
-					tmp_result, cnt);
+		if (iface == FI_HMEM_SYSTEM) {
+			ofi_atomic_swap_handler(op, datatype, dst, src, cmp, tmp_result, cnt);
+		} else {
+			char tmp_dst[SMR_INJECT_SIZE];
+			ofi_atomic_swap_handler(op, datatype, tmp_dst, src, cmp, tmp_result, cnt);
```

Comment:
How does this work?  tmp_dst in all of these cases is uninitialized.

Reply:
Good Catch... passes fi_rdm_atomic b/c it currently does no data validation.  Luke has a branch that is adding data validation to the test.  I will re-comment on this once I have successfully ran shm with his verification.

Reply:
Try using ubertest.  That should include data verification checks

Reply:
@shefty I commented out the entire `smr_do_atomic()` method and ran fi_ubertest for SHM, and all 774 test passed.  I ran like this:
```
# client
./fi_ubertest -u fabtests/test_configs/shm/all.test 127.0.0.1
# server
./fi_ubertest
```

This means that fi_ubertest must not be doing any data validation checks.

Reply:
@a-szegel The all.test config file does not include the verfication tests - this is done in FT_TEST_UNIT (as opposed to latency/bandwidth). Try running with verify.test instead

Reply:
I ran with @lrbison's validation tests and it "seems" to be working.  Luke's new tests still have some intermittent failure problems (which seem to be fabtest related).  They validate data on each iteration, and 1 iteration fails per 1000 iterations on most tests (consistently).

Reply:
@aingerson Ran verify.test for fi_uber shm and 100% of tests passed.  I didn't see any FI_HMEM in the output, so I am guessing I need to modify the config in someway to get FI_HMEM to run.

Reply:
@a-szegel Thanks for testing! HMEM support is enabled with the following arguments
-H -> HMEM aware but use host memory
-D <iface> -> use iface for device allocation
-i <idx> -> which device to use for device allocation

For example to run CUDA device->host test, run the server and client as so:
fi_ubertest -D cuda
fi_ubertest -H -u <config>

Reply:
It seems like the fabtests have a bug.  `ft_fill_buf()` is being called before `tx_msg_buf` is being initialized, so we are hitting an assert before we even get into the testing.